src/FSharpVSPowerTools.Logic/QuickInfoMargin.fs

```diff
@@ -0,0 +1,105 @@
+﻿namespace FSharpVSPowerTools.QuickInfo
+
+open System
+open System.Text
+open Microsoft.VisualStudio.Text.Editor
+open Microsoft.VisualStudio.Text
+open Microsoft.VisualStudio.Shell.Interop
+open FSharpVSPowerTools
+open FSharpVSPowerTools.ProjectSystem
+open FSharpVSPowerTools.AsyncMaybe
+open FSharp.ViewModule
+open Microsoft.FSharp.Compiler
+
+type QuickInfoVisual = FsXaml.XAML<"QuickInfoMargin.xaml", ExposeNamedProperties=true>
+
+type QuickInfoViewModel() as self = 
+    inherit ViewModelBase()
+    let quickInfo = self.Factory.Backing(<@@ self.QuickInfo @@>, "")
+    member __.QuickInfo
+        with get () = quickInfo.Value
+        and set (v) = quickInfo.Value <- v
+
+type QuickInfoMargin (textDocument: ITextDocument,
+                      view: ITextView, 
+                      vsLanguageService: VSLanguageService, 
+                      serviceProvider: IServiceProvider,
+                      projectFactory: ProjectFactory) = 
+
+    let updateLock = obj()
+    let model = QuickInfoViewModel()
+    let visual = QuickInfoVisual()
+    do visual.Root.DataContext <- model
+    let buffer = view.TextBuffer
+
+    let updateError (errors: (FSharpErrorSeverity * string list) []) = lock updateLock <| fun () -> 
+        model.QuickInfo <-
+            errors 
+            |> Array.map (fun (severity, errors) -> 
+                let errors = List.map String.trim errors
+                let title = 
+                    match errors with
+                    | [_] -> sprintf "%+A" severity
+                    | _ -> sprintf "%+As (%d)" severity errors.Length
+                title + ": " + 
+                (match errors with 
+                 | [e] -> e
+                 | _ -> 
+                    errors 
+                    |> List.mapi (fun i e -> sprintf "%d. %s" (i + 1) e) 
+                    |> List.toArray 
+                    |> String.concat " "))
+            |> String.concat " "
+
+    let flattenLines (x: string) =
+        match x with
+        | null -> ""
+        | x ->
+            x 
+            |> String.split StringSplitOptions.RemoveEmptyEntries [|"\r\n"; "\r"; "\n"|]
+            |> Array.mapi (fun i x -> 
+                if i > 0 && x.Length > 0 && Char.IsUpper x.[0] then 
+                    ". " + (String.trim x) else " " + (String.trim x))
+            |> String.concat ""
+            |> fun x -> if x.[x.Length - 1] <> '.' then x + "." else x
```

Comment:
Could we guard for empty strings here? As it is, the code could throw exceptions.